internal/witness/cmd/feeder/main.go

```diff
@@ -23,140 +23,175 @@ import (
 	"errors"
 	"flag"
 	"fmt"
+	"io/ioutil"
 	"net/http"
 	"net/url"
 	"time"
 
 	"github.com/golang/glog"
 	ct "github.com/google/certificate-transparency-go"
 	"github.com/google/certificate-transparency-go/client"
-	wa "github.com/google/certificate-transparency-go/internal/witness/api"
 	wh "github.com/google/certificate-transparency-go/internal/witness/client/http"
-	"github.com/google/certificate-transparency-go/internal/witness/verifier"
 	"github.com/google/certificate-transparency-go/jsonclient"
+	"github.com/google/certificate-transparency-go/loglist2"
 )
 
 var (
-	logURL   = flag.String("log_url", "", "The endpoint of the log HTTP API")
-	logPK    = flag.String("log_pk", "", "The base64-encoded log public key")
+	logList  = flag.String("log_list_url", "https://www.gstatic.com/ct/log_list/v3/log_list.json", "The location of the log list")
 	witness  = flag.String("witness_url", "", "The endpoint of the witness HTTP API")
 	interval = flag.Duration("poll", 10*time.Second, "How quickly to poll the log to get updates")
 )
 
+// feeder contains a map from log IDs to logData and a witness client.
 type feeder struct {
-	logID string
-	wsth  *ct.SignedTreeHead
-	c     *client.LogClient
-	w     wh.Witness
+	logs map[string]logData
+	w    wh.Witness
 }
 
-func main() {
-	flag.Parse()
-	ctx := context.Background()
-	// Set up the witness client.
-	if *logPK == "" {
-		glog.Exit("--log_pk must not be empty")
+// logData contains the latest witnessed STH for a log and a log client.
+type logData struct {
+	desc   string
+	wsth   *ct.SignedTreeHead
+	client *client.LogClient
+}
+
+// setupLogs populates the feeder's logs map based on the logList.
+func populateLogs(logListURL string) (map[string]logData, error) {
+	resp, err := http.Get(logListURL)
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve log list: %v", err)
 	}
-	var w wh.Witness
-	pk, err := ct.PublicKeyFromB64(*logPK)
+	defer resp.Body.Close()
+	body, err := ioutil.ReadAll(resp.Body)
 	if err != nil {
-		glog.Exitf("Failed to create public key: %v", err)
+		return nil, fmt.Errorf("failed to read HTTP response: %v", err)
 	}
-	wv, err := verifier.NewWitnessVerifier(pk)
+	// Get data for all usable logs.
+	logList, err := loglist2.NewFromJSON(body)
 	if err != nil {
-		glog.Exitf("Failed to create witness signature verifier: %v", err)
+		return nil, fmt.Errorf("failed to parse JSON: %v", err)
+	}
+	logs := make(map[string]logData)
+	usable := logList.SelectByStatus([]loglist2.LogStatus{loglist2.UsableLogStatus})
+	for _, operator := range usable.Operators {
+		for _, log := range operator.Logs {
+			logID := base64.StdEncoding.EncodeToString(log.LogID)
```

Comment:
Can you use `base64.URLEncoding` instead to avoid the other escaping you need to do on this ID?

Reply:
It seems important for the feeder and witness to use the same logID string that's used elsewhere in the CT ecosystem, which means computing it this way.  If that thinking changes then we can definitely do this (or just use hex or some other URL-safe encoding).